### PR TITLE
feat!: add PR-driven release workflow and release config domain

### LINKED ISSUE #34: Built-in release-please replacement: PR-driven release workflow
## Summary

Give pubm a first-class, built-in alternative to `release-please`: a GitHub workflow + bot that opens and continuously updates a "release PR" on changes to `main`, computes the next version and changelog, and then — once the PR is merged — runs only the publish phase. Users shouldn't need to stitch release-please + changesets + pubm together; pubm should cover the full path.

## Motivation

Today, users who want an automated PR-based release flow reach for `release-please` or a custom changesets-based action alongside pubm. That creates:

- **Two sources of truth** about "what's the next version" (release-please vs pubm's internal version logic)
- **Config duplication** — one tool's config in `.release-please-config.json`, another in `pubm.config.ts`
- **No shared dry-run signal.** release-please's PR doesn't know whether pubm will actually succeed at publish time; failures only surface after merge.
- **Missing plugin semantics.** release-please has no idea about pubm's plugin system, so things like Homebrew formula updates and external version sync can't be previewed in the release PR.

pubm already knows version bump logic, changelog generation, plugin orchestration, and dry-run. It should own the PR loop too.

## Proposed shape

Two runtime moments: **"on push to `main`"** and **"on release-PR merge"**.

### On push to `main`

1. pubm workflow computes the pending changeset set since the last release
2. Opens a PR titled e.g. `chore(release): <summary>` (or updates the existing open one)
3. The PR body includes:
   - Proposed next version per package (with the selected bump rule)
   - Generated changelog diff
   - Plugin previews (e.g. "brew formula will update sha to …", "will sync version in …")
   - A fresh dry-run result comment (pubm runs dry-run and posts the outcome)
4. Each subsequent push to `main` updates the same PR: recomputes version/changelog, re-runs dry-run, updates body + comment

### Bump version selection

The PR needs a way to override the auto-computed bump. Support multiple mechanisms, pick the one with best UX first:

- **Slash commands in PR comments.** `/pubm bump minor`, `/pubm bump major`, `/pubm release-as 2.0.0-beta.1`
- **PR labels.** `release:patch`, `release:minor`, `release:major`, `release:prerelease`
- **Changeset-based** (status quo extended). Existing `.changeset/*.md` files remain the default source; the above are overrides.

### On release-PR merge

- Runs publish phase only (materially equivalent to today's `--mode ci --phase publish`, if not identical)
- Tags / GitHub Releases / registry uploads / plugin-driven assets, same as current CI publish

## Why this matters for pubm

- Collapses the release toolchain from two tools to one
- Gives changeset/bump UX that's aware of pubm's plugin graph
- Dry-run-in-PR materially reduces "merged, publish failed" incidents
- Makes the "approval" idea feasible at the team level — the PR itself becomes the approval surface

## Open questions

- **Monorepo shape.** Single release PR per repo, or per-package? release-please supports both; we should decide early. Per-repo is simpler; per-package maps better to our fixed/independent versioning.
- **Independent vs fixed versioning.** How does this interact with `packages/core`'s fixed/independent config?
- **Bump-source precedence.** If a PR has both a `release:minor` label and a `/pubm bump major` comment, which wins? (Last-wins via comment is probably most predictable.)
- **Dry-run cost.** Running full dry-run on every push could be heavy. Debounce, or only run dry-run on label/comment/manual trigger?
- **Relationship to existing `--mode ci --phase publish`.** Is the merge-time publish literally identical? If so, the merge hook is thin glue; document the equivalence explicitly rather than hiding it.
- **State storage.** Where does the bot store the "which PR is the current release PR" pointer? PR label, branch naming convention (`release-please/*`-style), or a file under `.github/`?
- **Fallback / non-GitHub.** GitLab / Gitea adoption is a later step; design the core around git operations + a small forge adapter.
- **Interaction with prereleases.** Snapshot / canary / beta flows need explicit thinking — these probably don't go through the release PR.

## References

- https://github.com/googleapis/release-please — reference implementation to beat on ergonomics
- pubm's existing `--mode ci --phase publish` path
- pubm changeset workflow (`bun run changesets:add`)
- Related: approval mode, rollback-as-JSON issues

### PR DESCRIPTION
**Pull request by syi0808**
## Summary

Closes #34

- Add the PR-driven release workflow path in core: release scope grouping, rolling release PR reconciliation metadata, version/changelog materialization, tag/publish/GitHub Release steps, and publish-from-merged-release-PR guardrails.
- Redesign public release configuration under `release`: `release.versioning`, `release.changesets`, `release.commits`, `release.changelog`, and `release.pullRequest` replace the old scattered top-level release knobs.
- Remove `releasePr.enabled`, `releasePr.dryRun`, and `grouping: "auto"`; Release PR behavior is enabled by the workflow/action invocation, dry-run validation always runs, and grouping inherits `release.versioning.mode` unless overridden.
- Add shared release analysis for changesets plus conventional commits, including unversioned-change reporting with the default `warn` policy and no fallback bump.
- Make CI planning, interactive accept paths, `pubm changesets version`, and Release PR preparation use the same plan builder so fixed/linked groups and uneven fixed versions behave consistently.

## Behavior

- Pushes to `main` can open or update a rolling release PR instead of publishing directly.
- Merging the release PR runs the publish workflow, creates tags, publishes packages, and creates the GitHub Release.
- Changesets and releasable conventional commits are always analyzed when present; changesets win for overlapping package recommendations.
- Unversioned-only changes do not create a release PR by default; the workflow reports a warning unless configured as `ignore` or `fail`.
- `release.pullRequest.grouping`, `fixed`, and `linked` inherit from `release.versioning` and can be overridden inside `release.pullRequest`.

## Stacked PRs

- Docs split out to #43.
- Companion action PR: syi0808/pubm-actions#1, latest tested SHA `294512d`.

## Test plan

- [x] `bun run format`
- [x] `bun run typecheck`
- [x] `bun run test`
- [x] `bun run coverage`
- [x] `cd ../pubm-actions && PUBM_CORE_ENTRY=/Users/sung-yein/Workspace/pubm-issue-34-release-workflow/packages/core/src/index.ts bun run build`
- [x] `cd ../pubm-actions && bun run typecheck`
- [x] `cd ../pubm-actions && /Users/sung-yein/.nvm/versions/node/v24.5.0/bin/node node_modules/vitest/vitest.mjs --run`
- [x] QA repository `pubm-org/release-workflow-rolling-qa-20260502233021`: pushed a changeset, opened release PR #4 for `0.1.3`, verified version/changelog/changeset consumption diff, merged it, publish workflow succeeded, and GitHub Release `v0.1.3` was created.

## Notes

- GitHub Actions currently emits a Node 20 deprecation annotation for `actions/checkout@v4` and `actions/setup-node@v4`; it did not fail the QA workflow.
- Local GitHub Packages version reads still require a token with `read:packages`; the package API check returned 403 locally, while the workflow publish job had `packages: write` and completed successfully.

### CODE REVIEW

**Comment by syi0808** [comment]
@cubic-dev-ai review

**Comment by syi0808** [comment]
@codex review

File: packages/core/src/release-analysis/analyze.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  ).analyze(sourceContext);
+  const commits = await new ConventionalCommitSource(
+    ctx.config.release?.commits?.types ?? {},
+  ).analyze(sourceContext);
+
+  return {
+    recommendations: mergeRecommendations([changesets, commits]),
+    unversionedChanges: analyzeUnversionedCommits(ctx),
+  };
+}
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/release-analysis/analyze.ts:47]: @coderabbitai Fixed. Added `packages/core/tests/unit/release-analysis/analyze.test.ts` covering merged changeset/commit recommendations, `resolveKey`, and unversioned commit reporting.

File: packages/core/src/utils/exec.ts [hunk 2]
```diff
@@ -79,27 +81,90 @@ export async function exec(
...
+    let output = "";
+    stream.on("data", (chunk) => {
+      const text = Buffer.isBuffer(chunk)
+        ? chunk.toString("utf8")
+        : String(chunk);
+      output += text;
+      onChunk?.(text);
+    });
+    stream.on("error", reject);
+    stream.on("end", () => resolve(output));
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/utils/exec.ts]: @coderabbitai Fixed. Node stdout/stderr reading now uses streaming `TextDecoder` with a flush on `end`, so split multibyte UTF-8 chunks are decoded correctly. Covered in `packages/core/tests/unit/utils/exec.test.ts`.

File: packages/core/src/utils/exec.ts [hunk 3]
```diff
@@ -79,27 +81,90 @@ export async function exec(
...
+  });
+
+  const [stdout, stderr] = await Promise.all([
+    readNodeStream(proc.stdout, options.onStdout),
+    readNodeStream(proc.stderr, options.onStderr),
+  ]);
+  const exitCode = await new Promise<number>((resolve, reject) => {
+    proc.on("error", reject);
+    proc.on("close", (code: number | null) => resolve(code ?? 0));
+  });
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/utils/exec.ts]: @coderabbitai Fixed. Node process completion listeners are registered before stream reads, and null exit codes from signal termination now resolve as non-zero so `throwOnError` can fail correctly. Covered in `packages/core/tests/unit/utils/exec.test.ts`.

File: packages/core/src/config/defaults.ts [hunk 4]
```diff
@@ -160,15 +192,83 @@ export async function resolveConfig(
...
+    versioning,
+    changesets: {
+      ...defaultRelease.changesets,
+      ...config?.changesets,
+    },
+    commits: {
+      ...defaultRelease.commits,
+      ...config?.commits,
+      types: config?.commits?.types ?? defaultRelease.commits.types,
+    },
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/config/defaults.ts:230]: @coderabbitai Fixed. `resolveReleaseConfig()` now clones `release.versioning.fixed/linked` and `release.commits.types` from either defaults or input, with regression coverage in `packages/core/tests/unit/config/defaults.test.ts`.

File: packages/core/src/workflow/release-utils/changeset-consume.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      rmSync(filePath);
+      result.deletedFiles.push(filePath);
+    } else {
+      writeFileSync(
+        filePath,
+        generateChangesetContent(remaining, changeset.summary),
+        "utf-8",
+      );
+      result.rewrittenFiles.push(filePath);
+    }
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/workflow/release-utils/changeset-consume.ts:74]: @coderabbitai Fixed. `consumeChangesetsForScope()` now normalizes `deletedFiles` and `rewrittenFiles` entries to forward slashes before returning them. Existing assertions were updated to lock that behavior.

File: packages/core/src/workflow/release-utils/scope.ts [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+      const key = packageKey(pkg);
+      const aliases = [key, pkg.path, pkg.name].filter(Boolean) as string[];
+      if (
+        aliases.some((alias) => alias === ref || micromatch.isMatch(alias, ref))
+      ) {
+        keys.add(key);
+      }
+    }
+  }
+  return [...keys];
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/workflow/release-utils/scope.ts:131]: @coderabbitai Fixed. Configured release scope refs and aliases are normalized to forward slashes before exact/glob matching, with a Windows-style glob regression test in `packages/core/tests/unit/workflow/release-utils/scope.test.ts`.

File: packages/core/src/workflow/release-utils/release-pr-overrides.ts [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+): ReleasePrOverride | undefined {
+  const configured = { ...DEFAULT_RELEASE_PR_BUMP_LABELS, ...bumpLabels };
+  const labelSet = new Set(labels);
+  const matches = (Object.keys(configured) as ReleasePrBumpOverride[]).filter(
+    (bump) => labelSet.has(configured[bump]),
+  );
+
+  if (matches.length === 0) return undefined;
+  return { source: "label", kind: "bump", bump: highestBump(matches) };
+}
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/workflow/release-utils/release-pr-overrides.ts:66]: @coderabbitai Fixed. Release PR bump label resolution now ranks stable bumps above prerelease (`major > minor > patch > prerelease`), with coverage for `release:prerelease` plus `release:patch`.

File: packages/core/src/workflow/release-pr.ts [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+}
+
+function filterConfigForPackageKeys(
+  config: ResolvedPubmConfig,
+  packageKeys: ReadonlySet<string>,
+): ResolvedPubmConfig {
+  return Object.freeze({
+    ...config,
+    packages: config.packages.filter((pkg) => packageKeys.has(packageKey(pkg))),
+  });
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/workflow/release-pr.ts:385]: @coderabbitai Fixed. Scoped release PR configs now prune `fixed`/`linked` metadata on both top-level and nested `release.versioning` / `release.pullRequest`, so scoped prepare/dry-run/publish flows keep `ctx.config` internally consistent. Covered in `packages/core/tests/unit/workflow/release-pr.test.ts`.

File: packages/core/tests/helpers/make-context.ts [hunk 9]
```diff
@@ -23,14 +22,84 @@ export function makeTestConfig(
...
+      fixed: release?.pullRequest?.fixed ?? versioning.fixed,
+      linked: release?.pullRequest?.linked ?? versioning.linked,
+    },
+  };
+  config.versioning = config.release.versioning.mode;
+  config.fixed = config.release.versioning.fixed;
+  config.linked = config.release.versioning.linked;
+  config.updateInternalDependencies =
+    config.release.versioning.updateInternalDependencies;
+  config.changelog = config.release.changelog;
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/tests/helpers/make-context.ts:116]: @coderabbitai Fixed. `makeTestConfig()` now carries legacy top-level release overrides into nested `release.versioning` / `release.changelog`, while preserving nested overrides when provided. Added coverage in `packages/core/tests/unit/workflow/release-pr.test.ts`.

File: packages/pubm/src/commands/version-cmd.ts [hunk 10]
```diff
@@ -209,6 +170,47 @@ export async function runVersionCommand(
...
+    fixedMode: boolean,
+  ): void => {
+    const pkg = config.packages.find(
+      (candidate) => packageKey(candidate) === pkgKey,
+    );
+    if (!pkg) return;
+    const bumpType =
+      (fixedMode ? fixedBump : bumpTypeFromVersions(pkg.version, newVersion)) ??
+      recommendationBumpForPackage(config, recommendations, pkgKey) ??
+      fixedBump;
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/pubm/src/commands/version-cmd.ts:213]: @coderabbitai Fixed. `runVersionCommand()` now passes group-adjusted recommendations into bump creation and prefers the recommendation/group bump before falling back to `semver.diff`. The regression test now uses a patch-level prerelease graduation where `semver.diff` would infer the wrong bump.

File: packages/pubm/src/cli.ts [hunk 11]
```diff
@@ -360,10 +249,10 @@ export function createProgram(config?: ResolvedPubmConfig): Command {
 
           if (phase === "publish") {
             ctx.runtime.versionPlan =
-              createVersionPlanFromManifests(resolvedConfig);
+              createVersionPlanFromManifestVersions(resolvedConfig);
           } else if (isCI && phase === undefined) {
             if (!nextVersion) {
-              await applyVersionSourcePlan(ctx, resolvedConfig);
+              await applyVersionSourcePlan(ctx);
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/pubm/src/cli.ts:256]: @coderabbitai Fixed. The publish phase now builds `ctx.runtime.versionPlan` from override-filtered `ctx.config`, not the original `resolvedConfig`. Added CLI coverage for registry-filtered publish planning.

File: packages/pubm/tests/contracts/cli/phase-option-contract.test.ts [hunk 12]
```diff
@@ -271,6 +269,48 @@ vi.mock("@pubm/core", () => {
...
+            ctx.config.packages.map((pkg: any) => [
+              mockState.packageKey(pkg),
+              version,
+            ]),
+          ),
+        };
+      } else {
+        ctx.runtime.versionPlan = { mode: "independent", packages };
+      }
+    }),
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/pubm/tests/contracts/cli/phase-option-contract.test.ts:338]: @coderabbitai Fixed. The CLI contract mocks now read `release.versioning.mode` first and fall back to legacy `config.versioning`, with a publish-phase scenario that would fail if the nested release config path were ignored.

File: packages/core/src/workflow/release-utils/release-pr-comments.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    typeof input.scope === "string" ? input.scope : input.scope.displayName;
+  const lines = [
+    RELEASE_PR_DRY_RUN_COMMENT_MARKER,
+    "### release and publish dry-run passed",
+    "",
+    `Scope: \`${scope}\``,
+  ];
+
+  if (input.runUrl) {
+    lines.push("", `CI run: [View workflow run](${input.runUrl})`);
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/workflow/release-utils/release-pr-comments.ts:24]: @coderabbitai Fixed. Dry-run passed comments now render scope values as safe markdown code spans after collapsing newlines, including scopes containing backticks. Added edge-case coverage in `packages/core/tests/unit/workflow/release-utils/release-pr-comments.test.ts`.

File: packages/pubm/tests/unit/cli.test.ts [hunk 14]
```diff
@@ -41,18 +62,85 @@ const {
...
+    }
+
+    const packages = new Map(
+      config.packages.map((pkg: any) => [packageKey(pkg), pkg.version]),
+    );
+    if (config.versioning === "independent") {
+      return { mode: "independent", packages };
+    }
+    return { mode: "fixed", version: [...packages.values()][0], packages };
+  });
```
Comment (syi0808) [inline packages/pubm/tests/unit/cli.test.ts:138]: @coderabbitai Fixed. The unit CLI mocks now prefer `release.versioning.mode` with legacy fallback, and the publish-phase test verifies nested independent mode wins over a conflicting top-level fixed value.

File: packages/pubm/src/commands/init-workflows.ts [hunk 15]
```diff
@@ -248,15 +339,188 @@ ${envBlock}
...
+
+  return `name: pubm Release PR
+
+on:
+  push:
+    branches:
+      - ${defaultBranch}
+  workflow_dispatch:
+  issue_comment:
+    types: [created]
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/pubm/src/commands/init-workflows.ts:361]: @chatgpt-codex-connector Fixed. The generated release PR workflow now adds a job-level guard so `issue_comment` runs only continue for PR comments: `github.event_name != 'issue_comment' || github.event.issue.pull_request`. Covered in `packages/pubm/tests/unit/commands/init-workflows.test.ts`.

File: packages/core/src/version-source/plan.ts [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+    const pkgKey = key?.includes("::")
+      ? key
+      : config.packages
+          .map((pkg) => packageKey(pkg))
+          .find((candidate) => candidate === key);
+    const resolvedKey =
+      pkgKey ??
+      config.packages
+        .filter((pkg) => pkg.path === rec.packagePath)
+        .map((pkg) => packageKey(pkg))[0];
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/version-source/plan.ts]: @chatgpt-codex-connector Fixed. Unqualified path recommendations now expand to every matching package key before fixed/linked grouping, so multi-ecosystem packages at the same path are all preserved. Added coverage in `packages/core/tests/unit/version-source/plan.test.ts`.

File: packages/core/src/release-analysis/analyze.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/release-analysis/analyze.ts:47]: @coderabbitai[bot] Verified. `packages/core/tests/unit/release-analysis/analyze.test.ts` is present on the current PR branch and covers merged changeset/conventional-commit recommendations plus unversioned commit reporting. `cd packages/core && bun vitest --run tests/unit/release-analysis/analyze.test.ts` passes.

File: packages/core/src/workflow/release-utils/release-pr-overrides.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/workflow/release-utils/release-pr-overrides.ts:66]: @coderabbitai[bot] Verified. `highestBump()` now ranks `prerelease: 0`, `patch: 1`, `minor: 2`, and `major: 3`, so descending sort selects `major` over `prerelease`. The regression coverage in `packages/core/tests/unit/workflow/release-utils/release-pr-overrides.test.ts` asserts `release:prerelease` plus `release:patch` resolves to `patch`; the focused override test passes.

File: packages/core/src/utils/exec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/utils/exec.ts]: @coderabbitai[bot] Verified. `readNodeProcess()` already registers the child completion promise before awaiting stdout/stderr streams, and `code ?? 1` maps signal termination to a non-zero exit. The regression coverage in `packages/core/tests/unit/utils/exec.test.ts` covers both the fast-close race and signal termination cases.

File: packages/core/tests/unit/utils/exec.test.ts [hunk 17]
```diff
@@ -127,6 +130,34 @@ describe("exec", () => {
...
+    } finally {
+      process.env.PATH = originalPath;
+    }
+
+    expect(spawn).toHaveBeenCalledWith(
+      ["echo"],
+      expect.objectContaining({
+        env: expect.objectContaining({
+          PATH: expect.stringMatching(/node_modules\/\.bin[:;]$/),
+        }),
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/tests/unit/utils/exec.test.ts:173]: @coderabbitai[bot] Fixed. PATH construction now uses `path.join()` and `path.delimiter` in the implementation, and the test expectations use the same platform-safe path construction instead of hardcoded separators. (`packages/core/src/utils/exec.ts:36`, `packages/core/tests/unit/utils/exec.test.ts:67`, `packages/core/tests/unit/utils/exec.test.ts:158`)

File: packages/core/tests/unit/utils/exec.test.ts [hunk 18]
```diff
@@ -127,6 +130,34 @@ describe("exec", () => {
...
+    vi.stubGlobal("Bun", { spawn });
+    delete process.env.PATH;
+
+    try {
+      const { exec } = await import("../../../src/utils/exec.js");
+
+      await exec("echo");
+    } finally {
+      process.env.PATH = originalPath;
+    }
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/tests/unit/utils/exec.test.ts:160]: @coderabbitai[bot] Fixed. The PATH-unset test now deletes `process.env.PATH` for the setup and restores it by deleting again when the original value was absent, avoiding the literal `undefined` env value. (`packages/core/tests/unit/utils/exec.test.ts:136`)

File: packages/core/src/version-source/plan.ts [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+  const plan = createVersionPlanFromRecommendations(
+    ctx.config,
+    recommendations,
+  );
+  if (!plan) return;
+
+  ctx.runtime.versionPlan = plan;
+  ctx.runtime.changesetConsumed = recommendations.some(
+    (rec) => rec.source === "changeset",
+  );
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/version-source/plan.ts:102]: @coderabbitai[bot] Fixed. `applyVersionSourcePlan()` now derives `changesetConsumed` from recommendations that actually matched into the selected plan, not from the raw recommendation list. Added regression coverage for a dropped changeset plus surviving commit recommendation. (`packages/core/src/version-source/plan.ts:91`, `packages/core/tests/contracts/version-source/version-source.contract.test.ts:785`)

File: packages/core/tests/unit/workflow/release-pr.test.ts [hunk 20]
```diff
@@ -0,0 +1,1316 @@
...
+  RELEASE_PR_RELEASE_NOTES_START_MARKER,
+  sameReleasePrScope,
+} from "../../../src/workflow/release-utils/release-pr-metadata.js";
+import type { ReleasePrScope } from "../../../src/workflow/release-utils/scope.js";
+import { makeTestConfig } from "../../helpers/make-context.js";
+
+const releasePrState = vi.hoisted(() => ({
+  gitCalls: [] as Array<{ name: string; args: unknown[] }>,
+  status: "M  packages/a/package.json\nA  packages/a/CHANGELOG.md",
+  diffOutput: "packages/a/package.json\n",
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/tests/unit/workflow/release-pr.test.ts]: @coderabbitai[bot] Fixed. The release PR workflow test now builds manifest and changelog paths with `path.join()` constants and reuses them in mock status output, mock diff output, mock `writeVersions()` results, and changed-file expectations. (`packages/core/tests/unit/workflow/release-pr.test.ts:32`, `packages/core/tests/unit/workflow/release-pr.test.ts:315`, `packages/core/tests/unit/workflow/release-pr.test.ts:685`)

File: packages/pubm/tests/unit/cli.test.ts [hunk 21]
```diff
@@ -41,18 +64,85 @@ const {
...
+    }
+
+    const packages = new Map(
+      config.packages.map((pkg: any) => [packageKey(pkg), pkg.version]),
+    );
+    if (versioningMode(config) === "independent") {
+      return { mode: "independent", packages };
+    }
+    return { mode: "fixed", version: [...packages.values()][0], packages };
+  });
```
Comment (syi0808) [inline packages/pubm/tests/unit/cli.test.ts:138]: @coderabbitai[bot] Fixed. The CLI planner mocks now keep fixed-mode plans fixed for one-package and multi-package cases, compute the shared fixed version from the highest valid configured package version, and mark changesets consumed only from matched recommendations. I also aligned the real manifest planner and the CLI contract mock with the same fixed-mode behavior. (`packages/pubm/tests/unit/cli.test.ts:107`, `packages/pubm/tests/unit/cli.test.ts:170`, `packages/pubm/tests/unit/cli.test.ts:695`, `packages/core/src/workflow/release-pr.ts:274`)

File: packages/core/src/workflow/release-pr.ts [hunk 22]
```diff
@@ -0,0 +1,692 @@
...
+    const pkg = config.packages[0];
+    return {
+      mode: "single",
+      version: pkg?.version ?? "",
+      packageKey: pkg ? packageKey(pkg) : ".",
+    };
+  }
+
+  return { mode: "independent", packages };
+}
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/src/workflow/release-pr.ts:302]: @coderabbitai Fixed. `createVersionPlanFromManifestVersions()` now uses `config.release.versioning.mode ?? config.versioning` before deciding fixed mode, so publish-phase manifest planning honors the nested release config while keeping the legacy fallback. Added regression coverage for nested fixed mode, legacy fallback, and invalid manifest versions in `packages/core/tests/unit/workflow/release-pr.test.ts`.

File: packages/core/tests/unit/utils/exec.test.ts [hunk 23]
```diff
@@ -23,6 +26,7 @@ describe("exec", () => {
 
   afterEach(() => {
     vi.unstubAllGlobals();
+    vi.doUnmock("node:child_process");
     global.Bun = originalBun;
   });
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/tests/unit/utils/exec.test.ts:35]: @coderabbitai Fixed. The Bun global teardown now deletes `globalThis.Bun` when it was originally absent and restores it only when it existed, with the same cleanup applied to `exec`, `clipboard`, `open-url`, `spawn-interactive`, and asset pipeline tests.

File: packages/core/tests/unit/workflow/release-pr.test.ts [hunk 24]
```diff
@@ -0,0 +1,1346 @@
...
+const packageBPath = path.join("packages", "b").replace(/\\/g, "/");
+const packageCPath = path.join("packages", "c").replace(/\\/g, "/");
+const packageAManifestPath = path.join("packages", "a", "package.json");
+const packageBManifestPath = path.join("packages", "b", "package.json");
+const packageAChangelogPath = path.join("packages", "a", "CHANGELOG.md");
+const packageAStatusOutput = [
+  `M  ${packageAManifestPath}`,
+  `A  ${packageAChangelogPath}`,
+].join("\n");
+const gitDiffOutput = (...files: string[]) => `${files.join("\n")}\n`;
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/tests/unit/workflow/release-pr.test.ts]: @coderabbitai Fixed. `gitDiffOutput()` now normalizes mock `git diff --name-only` paths to forward slashes before joining, matching the manifest-file comparison path format in `findChangedVersionPackageKeys()`. (`packages/core/tests/unit/workflow/release-pr.test.ts:42`)

File: packages/core/tests/contracts/version-source/version-source.contract.test.ts [hunk 25]
```diff
@@ -118,16 +158,21 @@ describe("version source contracts", () => {
         "fix(core): commit-only fix",
         "COMMIT_FILES",
         "packages/core/src/index.ts",
+        "COMMIT_START def5678",
+        "fix(cli): repair flag parsing",
+        "COMMIT_FILES",
+        "packages/cli/src/cli.ts",
```
[the remark this answers is not in the document]
Comment (syi0808) [reply packages/core/tests/contracts/version-source/version-source.contract.test.ts]: @coderabbitai Fixed. The reviewed contract fixture now builds package paths and commit file paths through `path.join()` plus forward-slash normalization, and expected JS package keys are derived from the normalized path. (`packages/core/tests/contracts/version-source/version-source.contract.test.ts:34`)

**Comment by syi0808** [comment]
@cubic-dev-ai review

**Comment by syi0808** [comment]
@codex review

**Comment by syi0808** [comment]
@coderabbitai review

**Comment by syi0808** [comment]
@coderabbitai review
